staging/src/k8s.io/apiserver/pkg/util/logs/logs.go

```diff
@@ -77,3 +87,75 @@ func GlogSetter(val string) (string, error) {
 	}
 	return fmt.Sprintf("successfully set glog.logging.verbosity to %s", val), nil
 }
+
+var (
+	outputFile *os.File
+	logTeeWG   sync.WaitGroup
+)
+
+// InitLogFile sets up the log file, and redirects stdout & stderr to the file. This must be called
+// before any logging, otherwise those log lines will not be captured.
+func initLogFile() {
+	if !flag.Parsed() {
+		os.Stderr.WriteString("ERROR: InitLogFile before flag.Parse")
+		return
+	}
+
+	if *logFile == "" {
+		// If log-file is unset, nothing to do.
+		return
+	}
+
+	// If log-file is set, glog's logtostderr MUST be set to capture logs.
+	if f := flag.Lookup("logtostderr"); f != nil {
+		if f.Value.String() != "true" {
+			defer glog.Warningf("Forcing --logtostderr to true") // Defer to capture logs to file.
+			if err := f.Value.Set("true"); err != nil {
+				glog.Errorf("Failed to set --logstderr=true: %v", err)
+			}
+		}
+	}
+
+	// Setup the log file.
+	file, err := os.OpenFile(*logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		glog.Errorf("Failed to open %s for logging: %v", *logFile, err)
+		return
+	}
+	outputFile = file
+
+	if *teeLogs {
+		stdout := os.Stdout
+		pipeR, pipeW, err := os.Pipe()
+		if err != nil {
+			glog.Errorf("failed to create stdout pipe: %v", err)
+		}
+
+		// Since tee causes log writing to be asynchronous, we combine stdout & stderr into a single
+		// stream (tee'd to stdout) to maintain ordering.
+		os.Stdout = pipeW
+		os.Stderr = pipeW
+
+		logTeeWG.Add(1)
+		go func() {
+			io.Copy(io.MultiWriter(stdout, file), pipeR)
+			logTeeWG.Done()
+		}()
+	} else {
+		os.Stdout = file
```

Comment:
Why both stdout and stderr?

Reply:
The goal is to capture all the output to a file. In practice, we probably just need stderr (that's what glog writes to), but I think it would still be desireable to capture anything written to stdout. Also, the manifests for these components currently redirect both (e.g. https://github.com/kubernetes/kubernetes/blob/b6fd5d93b2048ae19cbcf21bafdd5cd7249d1f33/cluster/gce/manifests/kube-apiserver.manifest#L30)